### PR TITLE
feat(mandates): recurring payment support for bank redirect and bank debit payment method for stripe

### PR DESCRIPTION
**Pull request by hrithikesh026**
## Type of Change

- [ ] Bugfix
- [X] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

* Added support for bank debit recurring payment.
* Stored payment method data for bank debit in the mandate table's metadata column.
* Made changes in structure of stripes `PaymentIntentRequest` fields to support the new requirement

I have merged https://github.com/juspay/hyperswitch/pull/1231 to this branch since the changes were interdependent.
So merging this branch would automatically add features of https://github.com/juspay/hyperswitch/pull/1231

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/stripe/transformers.rs [hunk 1]
```diff
@@ -64,19 +64,32 @@ pub enum Auth3ds {
 }
 
 #[derive(Debug, Eq, PartialEq, Serialize)]
-#[serde(rename_all = "snake_case")]
+#[serde(rename_all = "snake_case", untagged)]
```
Comment (Narayanbhat166) [inline crates/router/src/connector/stripe/transformers.rs:67]: I guess you can use internally tagged enum type here https://serde.rs/enum-representations.html#internally-tagged

File: crates/router/src/connector/stripe/transformers.rs [hunk 2]
```diff
@@ -276,20 +289,32 @@ pub enum BankDebitData {
...
     pub bank_specific_data: BankDebitData,
 }
 
+#[derive(Debug, Eq, PartialEq, Serialize)]
+pub struct StripeBankDebitType {
+    #[serde(rename = "payment_method_types[]")]
+    pub payment_method_types: StripePaymentMethodType,
+}
+#[derive(Debug, Eq, PartialEq, Serialize)]
+#[serde(untagged)]
```
Comment (Narayanbhat166) [inline crates/router/src/connector/stripe/transformers.rs:304]: here you can use adjacent tagged enum 
https://serde.rs/enum-representations.html#adjacently-tagged

Comment (hrithikesh026) [reply crates/router/src/connector/stripe/transformers.rs:310]: @Narayanbhat166 changes made and pushed

**Review by jarnura — DISMISSED** [review]

File: crates/router/src/connector/stripe/transformers.rs [hunk 3]
```diff
@@ -1273,11 +1350,23 @@ impl TryFrom<&types::PaymentsAuthorizeRouterData> for PaymentIntentRequest {
...
                 }
             }
         };
 
+        let payment_method_types = if payment_method.is_some() {
+            //if recurring payment get payment_method_type
+            get_payment_method_type_for_saved_payment_method_payment(item).map(Some)?
+        } else {
+            payment_method_types
+        };
```
Comment (ArjunKarthik) [inline crates/router/src/connector/stripe/transformers.rs:1369]: Please add comments

File: crates/router/src/connector/stripe/transformers.rs [hunk 4]
```diff
@@ -1345,10 +1471,52 @@ impl TryFrom<&types::PaymentsAuthorizeRouterData> for PaymentIntentRequest {
...
+                Some(enums::PaymentMethodType::Eps)
+                | Some(enums::PaymentMethodType::Giropay)
+                | Some(enums::PaymentMethodType::Przelewy24) => {
+                    Err(errors::ConnectorError::NotSupported {
+                        message: "Eps, Giropay, Przelewy24 Recurring payments ".into(),
+                        connector: "Stripe",
+                        payment_experience: "Recurring Payments".into(),
+                    }
+                    .into())
+                }
```
Comment (ArjunKarthik) [inline crates/router/src/connector/stripe/transformers.rs:1494]:
```suggestion
```

This is not required, Is already handled in `StripePaymentMethodType::try_from(payment_method_type)`

File: crates/router/src/connector/stripe/transformers.rs [hunk 5]
```diff
@@ -1345,10 +1471,52 @@ impl TryFrom<&types::PaymentsAuthorizeRouterData> for PaymentIntentRequest {
...
+                Some(payment_method_type) => StripePaymentMethodType::try_from(payment_method_type),
+                None => Err(errors::ConnectorError::NotSupported {
+                    message: "payment_method_type for Recurring payments unavailable ".into(),
+                    connector: "Stripe",
+                    payment_experience: "Recurring Payments".into(),
+                }
```
Comment (ArjunKarthik) [inline crates/router/src/connector/stripe/transformers.rs:1500]: Throw missing required field

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

**Review by jarnura — DISMISSED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Review by ArjunKarthik — DISMISSED** [review]

File: crates/router/src/core/payments/helpers.rs [hunk 6]
```diff
@@ -1661,8 +1685,9 @@ pub fn generate_mandate(
                             .set_mandate_amount(Some(data.amount))
                             .set_mandate_currency(Some(data.currency))
                             .set_start_date(data.start_date)
-                            .set_end_date(data.end_date)
-                            .set_metadata(data.metadata),
+                            .set_end_date(data.end_date),
+                        // .set_metadata(data.metadata),
+                        // we are storing PaymentMethodData in metadata of mandate
```
Comment (SanchithHegde) [inline crates/router/src/core/payments/helpers.rs:1690]: Then do you need the `set_metadata()` call here or not?

**Review by SanchithHegde — APPROVED** [review]

**Review by ArjunKarthik — DISMISSED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Pull request merged**